File: Leetcode/202 - Happy Number/solution.js
var isHappy = function (n) {
    const getDigitSquareSum = (num) => {
        let sum = 0;
        while (num > 0) {
            const digit = num % 10;
            sum += digit * digit;
            num = Math.floor(num / 10);
        }
        return sum;
    };

    const numSet = new Set();
    while (!numSet.has(n)) {
        numSet.add(n);
        n = getDigitSquareSum(n);
        if (n === 1) return true;
    }

    return false;
};

File: Leetcode/202 - Happy Number/solution.py
def isHappy(self, n: int) -> bool:
    def get_digit_square_sum(num: int) -> int:
        total = 0
        while num > 0:
            digit = num % 10
            total += digit * digit
            num //= 10
        return total

    num_set: set[int] = set()
    while n not in num_set:
        num_set.add(n)
        n = get_digit_square_sum(n)
        if n == 1:
            return True

    return False